File: java/src/main/java/co/rerout/sdk/model/AbVariantInput.java
/*
 * rerout-java — Official Java SDK for the Rerout branded-link API.
 *
 * Copyright (c) 2026 Codecraft Solutions
 * Licensed under the MIT License — https://codecraftsolutions.co.za
 */

package co.rerout.sdk.model;

import com.google.gson.annotations.SerializedName;

/**
 * A request item for an A/B test variant on a Smart Link. Used in the
 * {@code abVariants} list of {@link CreateLinkInput} and
 * {@link UpdateLinkInput}.
 *
 * <p>Unlike {@link AbVariant} (the response shape) there is no {@code id} — the
 * server assigns one — and {@code weight} is optional; leave it {@code null} to
 * take the server default. The field names mirror the server shape so Gson
 * serializes it directly, omitting a {@code null} weight.
 */
public final class AbVariantInput {

    @SerializedName("target_url")
    private final String targetUrl;

    private final Integer weight;

    /**
     * Creates an {@code AbVariantInput} with the server-default weight.
     *
     * @param targetUrl the destination this variant resolves to
     */
    public AbVariantInput(String targetUrl) {
        this(targetUrl, null);
    }

    /**
     * Creates an {@code AbVariantInput}.
     *
     * @param targetUrl the destination this variant resolves to
     * @param weight    the relative traffic weight, or {@code null} for the
     *                  server default
     */
    public AbVariantInput(String targetUrl, Integer weight) {
        this.targetUrl = targetUrl;
        this.weight = weight;
    }

    /** {@return the destination this variant resolves to} */
    public String getTargetUrl() {
        return targetUrl;
    }

    /** {@return the relative traffic weight, or {@code null} for the default} */
    public Integer getWeight() {
        return weight;
    }
}

File: php/src/Models/AbVariantInput.php
<?php

declare(strict_types=1);

namespace Rerout\Models;

/**
 * A request item for an A/B test variant on a Smart Link. Used in the
 * `ab_variants` array of {@see CreateLinkInput} and {@see UpdateLinkInput}.
 *
 * Unlike {@see AbVariant} (the response shape) there is no `id` — the server
 * assigns one — and `weight` is optional; omit it to take the server default.
 */
final readonly class AbVariantInput
{
    /**
     * @param string   $targetUrl Destination this variant resolves to.
     * @param int|null  $weight    Relative traffic weight. Omit for the server default.
     */
    public function __construct(
        public string $targetUrl,
        public ?int $weight = null,
    ) {
    }

    /**
     * Render to the JSON-serialisable shape the API expects. A null `weight`
     * is stripped so the server applies its default.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $out = ['target_url' => $this->targetUrl];
        if ($this->weight !== null) {
            $out['weight'] = $this->weight;
        }

        return $out;
    }
}
